client/src/main/java/org/apache/rocketmq/client/trace/AsyncTraceDispatcher.java

```diff
@@ -243,113 +247,133 @@ class AsyncRunnable implements Runnable {
-                if (transBeanList == null) {
-                    transBeanList = new ArrayList<TraceTransferBean>();
-                    transBeanMap.put(key, transBeanList);
-                }
-                TraceTransferBean traceData = TraceDataEncoder.encoderFromContextBean(context);
-                transBeanList.add(traceData);
-            }
-            for (Map.Entry<String, List<TraceTransferBean>> entry : transBeanMap.entrySet()) {
-                String[] key = entry.getKey().split(String.valueOf(TraceConstants.CONTENT_SPLITOR));
-                String dataTopic = entry.getKey();
-                String regionId = null;
-                if (key.length > 1) {
-                    dataTopic = key[0];
-                    regionId = key[1];
-                }
-                flushData(entry.getValue(), dataTopic, regionId);
+        public void addTraceTransferBean(TraceTransferBean traceTransferBean) {
+            initFirstBeanAddTime();
+            this.traceTransferBeanList.add(traceTransferBean);
+            this.currentMsgSize += traceTransferBean.getTransData().length();
+            if (currentMsgSize >= maxMsgSize) {
+                List<TraceTransferBean> dataToSend = new ArrayList(traceTransferBeanList);
+                AsyncDataSendTask asyncDataSendTask = new AsyncDataSendTask(regionId, dataToSend);
+                traceExecutor.submit(asyncDataSendTask);
+
+                this.clear();
             }
         }
 
-        /**
-         * Batch sending data actually
-         */
-        private void flushData(List<TraceTransferBean> transBeanList, String dataTopic, String regionId) {
-            if (transBeanList.size() == 0) {
+        public void sendAllData() {
+            if (this.traceTransferBeanList.isEmpty()) {
                 return;
             }
-            // Temporary buffer
+            List<TraceTransferBean> dataToSend = new ArrayList(traceTransferBeanList);
+            AsyncDataSendTask asyncDataSendTask = new AsyncDataSendTask(regionId, dataToSend);
+            traceExecutor.submit(asyncDataSendTask);
+
+            this.clear();
+        }
+
+        private void initFirstBeanAddTime() {
+            if (firstBeanAddTime == 0) {
+                firstBeanAddTime = System.currentTimeMillis();
+            }
+        }
+
+        private void clear() {
+            this.firstBeanAddTime = 0;
+            this.currentMsgSize = 0;
+            this.traceTransferBeanList.clear();
+        }
+    }
+
+
+    class AsyncDataSendTask implements Runnable {
+        public final String regionId;
+        public final List<TraceTransferBean> traceTransferBeanList;
+
+        public AsyncDataSendTask(String regionId, List<TraceTransferBean> traceTransferBeanList) {
+            this.regionId = regionId;
+            this.traceTransferBeanList = traceTransferBeanList;
+        }
+
+        @Override
+        public void run() {
             StringBuilder buffer = new StringBuilder(1024);
```

Comment:
There is a `try catch` in the following invoked method `sendTraceDataByMQ()`, is this enough?

<img width="760" alt="image" src="https://user-images.githubusercontent.com/18216266/164983447-e3739a8d-4741-4b53-bd8d-44a555127475.png">

Reply:
Currently, it is enough since the code outside the try-catch is simple, but it may introduce vulnerabilities if someone else adds code to it in the future.
It is better to show the best practice if you have spare time.